src/Data/Functor/Foldable.hs

```diff
@@ -849,3 +1114,12 @@ instance (GCoerce f g, GCoerce f' g') => GCoerce (f :*: f') (g :*: g') where
 instance (GCoerce f g, GCoerce f' g') => GCoerce (f :+: f') (g :+: g') where
     gcoerce (L1 x) = L1 (gcoerce x)
     gcoerce (R1 x) = R1 (gcoerce x)
+
+-- $intro
+-- TODO
+
+-- $patternFunctors
+-- TODO
```

Comment:
I prefer `yaya`'s term of "pattern functor" rather than "base" functor. How do the authors feel about this? Either way, this section should explain what base/pattern functors are.

Reply:
The code has `Base`, and I'd avoid using many names for same concept.

Also `pattern` is so overloaded: Meijer 1991 calls recursion-schemes recursion patterns, we have pattern matching, ..

Reply:
You're right, calling them pattern functors and then having the connection made with `Base` is even more confusing. I don't like the name `Base` but it's probably not worth the breaking API change now. Will change this to "base functors".

Reply:
> The code has Base, and I'd avoid using many names for same concept

When we make the big breaking gather-scatter change, maybe we can include a few other breaking name changes?

Reply:
If we do, here are the names I would like to suggest: "recursive type" for `[a]`, "non-recursive type" or "single step type" for `ListF a r` (and I'd like to  use `r` for recursive positions instead of whatever's the next available letter, in this case `b`), "recursive function" for `cata f`, and "non-recursive function" for `f` itself. Having a consistent name for `f` across recursion schemes would be especially useful for writing the part of the documentation which talks about multiple recursion schemes, as the current documentation uses a slightly different name in each one: _f_-algebra, _f_-coalgebra, _f_-w-algebra...

Reply:
If people haven't already seen it, I'd suggest comparing against `yaya`, which I think has some slightly better terminology: http://hackage.haskell.org/package/yaya-0.2.1.0/docs/Yaya-Fold.html. Because `yaya` is using functional dependencies, there's no need to "name" what the functor is, but it does use the concept of `Steppable` (though that's for `Corecursive`). Note also that @sellout exports `Algebra` type aliases, which we may want to consider.

Reply:
The `Algebra` type-aliases are two-edged sword. Makes type-signatures nicer, but opaque.

> and I'd like to use `r` for recursive positions instead of whatever's the next available letter,

That's not breaking change, we are free to do alpha-conversions on will!

Reply:
> Because yaya is using functional dependencies, there's no need to "name" what the functor is.

You still need to talk about "base functor" related to the recursive structure.

*single step type of `[a]` is `ListF a`* is IMHO as arbitrary as *base functor of [a] is `ListF a`*. I also like that *functor* is mentioned.

Reply:
@ocharles Thanks for the mention, and I appreciate the support. My hope is to eventually get yaya merged into recursion-schemes, rather than splitting the user base, but currently I am experimenting with different approaches to things, many of which are … uncertain.

To address the points here, I think I may agree that “base functor” is a better name than “pattern functor”, and I try to call out the other existing differences with recursion-schemes and their justifications in https://github.com/sellout/yaya#differences-from-recursion-schemes (although that’s at least _slightly_ out of date).

I’m also working on a book (https://github.com/vil1/recursion-schemes-cookbook) intended to explain “best practices” for recursion schemes. There’s a WIP (unpushed) section in that that discusses a problem with `Algebra` type aliases. I’ve stopped recommending their usage, primarily because they encourage users to write F-algebras directly. Often, the more natural (and more specifically typed) definition is something like
```haskell
foo :: Natural -> XNor a Text -> Either Error Text
```
which can then have “algebra transformers” applied to it, which transform it into an F-algebra, like
```haskell
-- | :: Compose ((,) Natural) (XNor a) (Either Error Text) -> Either Error Text
foo' :: Algebra (Compose ((,) Natural) (XNor a)) (Either Error Text)
foo' = lowerElgotAlgebra (uncurry foo <=< sequenceA)
```
(And definitely open to other contributors to the book.)

While were here, I guess … one deeper frustration I have with recursion-schemes is the type class arrangement
```haskell
class Recursive t where
  project :: t -> Base t t
  cata :: (Base t a -> a) -> t -> a
``` is backward. It should either split it as yaya does, with `project` and `embed` in a separate type class (with an isomorphism law) or
```haskell
class Recursive t where
  embed :: Base t t -> t
  cata :: (Base t a -> a) -> t -> a
``` which allows for the `cataRefl = uncurry (==) . (cata embed && id)` law. The problem with the current classes is apparent in `gcata distPara` (and `gana distApo`), which require both `Recursive` and `Corecursive` instances for a type (which is generally unsafe to have).

Reply:
`yaya` is `AGPL`, you don't simply merge it in. (You already have a contributor which owns copyright on its own code; especially if you get significant contribution to `yaya`, it will be a problem).

Reply:
I think we need explicit examples of

>>> which require both Recursive and Corecursive instances for a type (**which is generally unsafe to have**)

Reply:
I’m aware of the licensing issues involved, no worries.

The general unsafety of folding lazy (corecursive) structures and unfolding to strict structures is pretty well established. E.g.,
```haskell
let nats = 0 : fmap (+ 1) nats
in foldr (+) 0 nats
```
won’t terminate. In general, only `Base t ~ Const t` has safe instances of both. The separation of safe and unsafe instances is probably the primary motivating force behind yaya.

Reply:
But let me phrase that frustration a different way – I don’t think there are any laws that relate `project` to `cata`, but there _are_ laws for `cata` on its own, as well as laws that relate `project` to `embed` and laws that relate `embed` to `cata`. So the `project` in `Recursive` is lawless, whereas `embed` in `Recursive` wouldn’t be.

Reply:
@sellout so you argue that having `Projectable` and `Steppable` (Why not Embeddable?) for `data Stream a = a :> Stream a` but we shouldn't have `Recursive` for it?

Reply:
Yeah, so as I mentioned, yaya is still experimenting with a bunch of stuff, and the type class structure is part of that. The law-introducing classes are
- cata
- ana
- embed + project
- embed + cata
- project + ana

Ideally, I would like to avoid having any lawless type classes, so I had provided
- cata (Recursive)
- ana (Corecursive)
- embed + project (Steppable)

as three separate classes. But, I’ve run into cases where `embed` is impossible to implement, so I ended up pulling out `Projectable` so I could still implement `project` in that case. However, `Projectable` is lawless. You need to either implement `Steppable` or `Corecursive` to introduce laws on it. So I’d rather not have it exist. So it might be better to move to the
- embed + cata
- project + ana

classes. Still waffling. I’d kind of like to split each operation into its own class, if I could somehow require that the lawless classes had one of their “companion” classes implemented.